server.js

```diff
@@ -1632,6 +1625,64 @@ cache(function (data, match, sendBadge, request) {
   });
 }));
 
+// npm weekly download integration.
+camp.route(/^\/npm\/dw\/(.*)\.(svg|png|gif|jpg|json)$/,
+cache(function (data, match, sendBadge, request) {
+  var pkg = encodeURIComponent(match[1]);  // eg, "express" or "@user/express"
+  var format = match[2];
+  var apiUrl = 'https://api.npmjs.org/downloads/point/last-week/' + pkg;
+  var badgeData = getBadgeData('downloads', data);
+  request(apiUrl, function (err, res, buffer) {
+    if (err != null) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+    try {
+      var weekly = JSON.parse(buffer).downloads || 0;
+      badgeData.text[1] = metric(weekly) + '/week';
+      if (weekly === 0) {
+        badgeData.colorscheme = 'red';
+      } else {
+        badgeData.colorscheme = 'brightgreen';
+      }
+      sendBadge(format, badgeData);
+    } catch(e) {
+      badgeData.text[1] = 'invalid';
+      sendBadge(format, badgeData);
+    }
+  });
+}));
+
+// npm yearly download integration
+camp.route(/^\/npm\/dy\/(.*)\.(svg|png|gif|jpg|json)$/,
+cache(function (data, match, sendBadge, request) {
+  var pkg = encodeURIComponent(match[1]);  // eg, "express" or "@user/express"
+  var format = match[2];
+  var apiUrl = 'https://api.npmjs.org/downloads/point/last-year/' + pkg;
+  var badgeData = getBadgeData('downloads', data);
+  request(apiUrl, function (err, res, buffer) {
+    if (err != null) {
+      badgeData.text[1] = 'inaccessible';
+      sendBadge(format, badgeData);
+      return;
+    }
+    try {
+      var yearly = JSON.parse(buffer).downloads || 0;
+      badgeData.text[1] = metric(yearly) + '/year';
+      if (yearly === 0) {
+        badgeData.colorscheme = 'red';
+      } else {
+        badgeData.colorscheme = 'brightgreen';
+      }
+      sendBadge(format, badgeData);
+    } catch(e) {
+      badgeData.text[1] = 'invalid';
+      sendBadge(format, badgeData);
+    }
+  });
+}));
```

Comment:
Most of this code is copy and pasted across all three... Could you please generalise the parts of the code that's reusable?